### PR TITLE
[token-detection-controller] Add tests for API changes from merge with `DetectTokensController`

### LINKED ISSUE #3626: Merge extension `DetectTokensController` into core `TokenDetectionController`
`TokenDetectionController` and `DetectTokensController` implement similar functionality, but have separate implementations. They should be merged into a single controller in core.

### PR DESCRIPTION
**Pull request by MajorLift**
## Explanation

- This PR adds unit tests for the token-detection-controller API changes implemented in #3775.
- The following changes are covered in the new tests:
  - Don't detect if keyring-controller `isUnlocked` state is false.
  - Subscribe to `KeyringController:unlock`, `KeyringController:lock` events
  - Subscribe to `AccountsController:selectedAccountChange` event
  - Detect tokens using `@metamask/contract-metadata` static token list if on mainnet and `useTokenDetection` is false.
  - Call `trackMetaMetricsEvent` for every detected token.
- The aim of this PR isn't to achieve 100% test coverage for token-detection-controller. That will be the goal of follow-up ticket #1615

## References

- Closes #3626 
- Follows from:
  - https://github.com/MetaMask/core/pull/3775/
  - https://github.com/MetaMask/core/pull/3690/
- Followed by #1615 
  
## Changelog

N/A

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Comment by mcmire** [comment]
I'll wait until the parent PR is merged before giving final approval on this PR.

File: packages/assets-controllers/src/TokenDetectionController.test.ts
Comment (mcmire) [inline packages/assets-controllers/src/TokenDetectionController.test.ts:525]: Nit: `false` is a boolean, not a string:

``` suggestion
    describe('when "disabled" is false', () => {

File: packages/assets-controllers/src/TokenDetectionController.test.ts
Comment (mcmire) [inline packages/assets-controllers/src/TokenDetectionController.test.ts:679]: Similar as previous nit:

``` suggestion
    describe('when "disabled" is true', () => {

File: packages/assets-controllers/src/TokenDetectionController.test.ts
Comment (mcmire) [inline packages/assets-controllers/src/TokenDetectionController.test.ts:1942]: I'm curious if adding a new option here could confuse people reading the tests into thinking that the controller takes this option. Should this be a separate option?

**Review by mcmire — COMMENTED** [review]
Aside from the nitpicks I had one question.

File: packages/assets-controllers/src/TokenDetectionController.test.ts
Comment (MajorLift) [reply packages/assets-controllers/src/TokenDetectionController.test.ts]: Fixed throughout test file: https://github.com/MetaMask/core/pull/3867/commits/ad3593df4a9fdb8c6b55380f6caeb6393adb107a

File: packages/assets-controllers/src/TokenDetectionController.test.ts
Comment (MajorLift) [reply packages/assets-controllers/src/TokenDetectionController.test.ts]: Same as above: https://github.com/MetaMask/core/pull/3867/commits/ad3593df4a9fdb8c6b55380f6caeb6393adb107a

File: packages/assets-controllers/src/TokenDetectionController.test.ts
Comment (MajorLift) [reply packages/assets-controllers/src/TokenDetectionController.test.ts]: I wasn't sure where to put `isKeyringUnlocked`, but I like the idea of keeping `options` limited to controller constructor. 

Applied here: https://github.com/MetaMask/core/pull/3867/commits/95b34ac7ab409587a1b998ef41db5a9d57d6bdad

**Review by mcmire — APPROVED** [review]
LGTM!

**Pull request merged**
